planning/obstacle_velocity_planner/src/node.cpp

```diff
@@ -0,0 +1,689 @@
+
+    // rough detection area filtering without polygons
+    const double dist_from_obstacle_to_traj = [&]() {
+      return tier4_autoware_utils::calcLateralOffset(decimated_traj.points, object_pose.position);
+    }();
+    if (dist_from_obstacle_to_traj > obstacle_filtering_param_.rough_detection_area_expand_width) {
+      RCLCPP_INFO_EXPRESSION(
+        get_logger(), is_showing_debug_info_,
+        "Ignore obstacles since it is far from the trajectory.");
+      continue;
+    }
+
+    // calculate collision points
+    const auto obstacle_polygon =
+      polygon_utils::convertObstacleToPolygon(object_pose, predicted_object.shape);
+    std::vector<geometry_msgs::msg::Point> collision_points;
+    const auto first_within_idx = polygon_utils::getFirstCollisionIndex(
+      decimated_traj_polygons, obstacle_polygon, collision_points);
+
+    // precide detection area filtering with polygons
+    geometry_msgs::msg::Point nearest_collision_point;
+    if (first_within_idx) {  // obsacles inside the trajectory
+      // calculate nearest collision point
+      nearest_collision_point =
+        calcNearestCollisionPoint(first_within_idx.get(), collision_points, decimated_traj);
+      debug_data.collision_points.push_back(nearest_collision_point);
+
+      const bool is_angle_aligned = isAngleAlignedWithTrajectory(
+        decimated_traj, object_pose,
+        obstacle_filtering_param_.crossing_obstacle_traj_angle_threshold);
+      const double has_high_speed =
+        std::abs(object_velocity) > obstacle_filtering_param_.min_obstacle_crossing_velocity;
+
+      // ignore running vehicle crossing the ego trajectory with high speed with some condition
+      if (!is_angle_aligned && has_high_speed) {
+        const double collision_time_margin = calcCollisionTimeMargin(
+          current_pose, current_vel, nearest_collision_point, predicted_object,
+          first_within_idx.get(), decimated_traj, decimated_traj_polygons);
+        if (collision_time_margin > obstacle_filtering_param_.margin_for_collision_time) {
+          // Ignore condition 1
+          // Ignore vehicle obstacles inside the trajectory, which is crossing the trajectory with
+          // high speed and does not collide with ego in a certain time.
+          RCLCPP_INFO_EXPRESSION(
+            get_logger(), is_showing_debug_info_,
+            "Ignore inside obstacles since it will not collide with the ego.");
+
+          debug_data.intentionally_ignored_obstacles.push_back(predicted_object);
+          continue;
+        }
+      }
+    } else {  // obstacles outside the trajectory
+      const double max_dist =
+        3.0;  // std::max(vehicle_info_.max_longitudinal_offset_m, vehicle_info_.rear_overhang) +
+      // std::max(shape.dimensions.x, shape.dimensions.y) / 2.0;
+      const auto predicted_path_with_highest_confidence =
+        getHighestConfidencePredictedPath(predicted_object);
```

Comment:
Note that using only a predicted path with highest confidence may not work well on branch roads, etc.